src/PaymentGateways/DataTransferObjects/GatewayPaymentData.php

```diff
@@ -79,12 +91,62 @@ public static function fromArray(array $array)
         $self->date = $array['date'];
         $self->gatewayId = $array['gatewayId'];
         $self->donationId = $array['donationId'];
+        $self->formId = $array['formId'];
+        $self->formTitle = $array['formTitle'];
         $self->purchaseKey = $array['purchaseKey'];
         $self->donorInfo = $array['donorInfo'];
         $self->cardInfo = $array['cardInfo'];
         $self->billingAddress = $array['billingAddress'];
         $self->redirectUrl = give_get_success_page_uri();
 
+        $self->setDonationTitle();
+
         return $self;
     }
+
+    /**
+     * @unlreased
+     *
+     * @param int|null $titleLength
+     *
+     * @return string
+     */
+    public function getDonationTitle($titleLength = null)
+    {
+        // Cut the length
+        if ( ! $titleLength) {
+            return $this->donationTitle;
+        }
+
+        return substr($this->donationTitle, 0, $titleLength);
+    }
+
+    /**
+     * Set donationTitle property.
+     *
+     * @unlreased
+     */
+    private function setDonationTitle()
+    {
+        $this->donationTitle = $this->formTitle;
+        $price_id = $this->priceId;
+
+        // Verify has variable prices.
+        if (give_has_variable_prices($this->formId)) {
+            $item_price_level_text = give_get_price_option_name($this->formId, $price_id, 0, false);
+
+            /**
+             * Output donation level text if:
+             *
+             * 1. It's not a custom amount
+             * 2. The level field has actual text and isn't the amount (which is already displayed on the receipt).
+             */
+            if ('custom' !== $price_id && ! empty($item_price_level_text)) {
+                // Matches a donation level - append level text.
+                $this->donationTitle .= " - $item_price_level_text";
+            }
+        }
+
+        // TODO: discuss backward compatibility about removed filter
+    }
 }
```

Comment:
@ravinder I know you're still working on this but at first glance I saw you added some form properties to `GatewayPaymentData`.  We really want to avoid putting the context of a `form` like `formId`, and `formTitle` inside this DTO.  Doing so limits the scope of the gateway which is our goal - so it's only aware of a donation/payment record - not necessarily the "GiveWP" form.  We can still use helpers inside our gateway specific needs like grabbing a form ID & title using the donation ID so I would just move this stuff into a repository & use the supplied `$paymentData->donationId` to get what you need.

Reply:
@jonwaldstein Yes, I agree with you. I was thinking the same when I saw this DTO. I was planning to discuss this yesterday but was struck with another issue. I will update it. 

FYI, I will keep functionality but I will grab required form data from helpers.